File: routes/prod/v2.js
const fs = require('fs')
const Security = require("../../models/security.models");
const ERRORS = require("../../error.js");
const FIND_USERS = require("../../models/find_users.models");
const FIND_PLATFORMS = require("../../models/find_platforms.models");
const STORE_TOKENS = require("../../models/store_tokens.models");
const VERIFY_USERS = require("../../models/verify_user.models");
const STORE_SESSION = require("../../models/store_sessions.models");
const FIND_USERSINFO = require("../../models/find_usersinfo.models");
const STORE_USERS = require("../../models/store_users.models");
const INIT_2FA = require("../../models/init_2fa.models");
const VERIFY_2FA = require("../../models/verify_2fa.models");
const VERIFY_SV = require("../../models/verify_sv.models");
const FIND_SESSION = require("../../models/find_sessions.models");
const UPDATE_SESSION = require("../../models/update_sessions.models");
const FIND_USERS_PLATFORMS = require("../../models/find_users_platform.models");

var rootCas = require('ssl-root-cas').create()

require('https').globalAgent.options.ca = rootCas

// ==================== PRODUCTION ====================
module.exports = (app, configs) => {
    let PLATFORMS = require("../../libs/platforms");

    if ((configs.hasOwnProperty("ssl_api") && configs.hasOwnProperty("PEM")) && fs.existsSync(configs.ssl_api.PEM)) {
        rootCas.addFile('/var/www/ssl/server.pem')
    }

    app.post("/signup", async (req, res, next) => {
        try {
            // ==================== REQUEST BODY CHECKS ====================
            if (!req.body.phone_number) {
                console.error("NO PHONE NUMBER");
                throw new ERRORS.BadRequest();
            };

            if (!req.body.name) {
                console.error("NO NAME");
                throw new ERRORS.BadRequest();
            };

            if (!req.body.country_code) {
                console.error("NO COUNTRY CODE");
                throw new ERRORS.BadRequest();
            };

            if (!req.body.password) {
                console.error("NO PASSWORD");
                throw new ERRORS.BadRequest();
            };

            // TODO ADD MIDDLEWARE CHECKS
            if (req.body.password.length < 8) {
                console.error("PASSWORD < 8 CHARS");
                throw new ERRORS.BadRequest();
            };
            // =============================================================
            const COUNTRY_CODE = req.body.country_code;
            const NAME = req.body.name;
            const PHONE_NUMBER = req.body.phone_number;
            const FULL_PHONE_NUMBER = req.body.country_code + req.body.phone_number;
            const PASSWORD = req.body.password;

            let userInfo = await FIND_USERSINFO(COUNTRY_CODE, PHONE_NUMBER);

            if (!userInfo) {
                let userId = await STORE_USERS(COUNTRY_CODE, PHONE_NUMBER, PASSWORD, NAME);
                let init_2fa = await INIT_2FA(userId, FULL_PHONE_NUMBER);

                return res.status(200).json({
                    session_id: init_2fa.session_id,
                    svid: init_2fa.svid
                })
            } else {
                console.error("USER ALREADY HAS A RECORD IN USERINFO TABLE");
                throw new ERRORS.Conflict();
            };
        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(404).send(err.message);
            } // 404

            console.error(err);
            return res.status(500).send("internal server error");
        }
    });

    app.put("/signup", async (req, res, next) => {
        try {
            // ==================== REQUEST BODY CHECKS ====================
            if (!req.body.code) {
                console.error("NO CODE");
                throw new ERRORS.BadRequest();
            };

            if (!req.body.session_id) {
                console.error("NO SESSION ID");
                throw new ERRORS.BadRequest();
            };

            if (!req.body.svid) {
                console.error("NO SVID");
                throw new ERRORS.BadRequest();
            };
            // =============================================================
            const CODE = req.body.code;
            const SESSION_ID = req.body.session_id;
            const SVID = req.body.svid;

            let {
                usersInfo,
                user
            } = await VERIFY_SV(SESSION_ID, SVID);
            let verify_2fa = await VERIFY_2FA(usersInfo.full_phone_number, CODE, SESSION_ID);

            if (verify_2fa) {
                var security = new Security(user.password);

                await usersInfo.update({
                    phone_number: security.encrypt(usersInfo.phone_number).e_info,
                    name: security.encrypt(usersInfo.name).e_info,
                    country_code: security.encrypt(usersInfo.country_code).e_info,
                    full_phone_number: security.hash(usersInfo.full_phone_number),
                    status: "verified",
                    iv: security.iv
                }).catch(error => {
                    console.error("ERROR UPDATING USER'S INFO AFTER SMS VERIFICATION");
                    throw new ERRORS.InternalServerError(error);
                });

                return res.status(200).json();
            };
        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(404).send(err.message);
            } // 404

            console.error(err);
            return res.status(500).send("internal server error");
        }
    });

    app.post("/login", async (req, res, next) => {
        try {
            // ==================== REQUEST BODY CHECKS ====================
            if (!req.body.phone_number) {
                console.error("NO PHONE NUMBER");
                throw new ERRORS.BadRequest();
            };

            if (!req.body.password) {
                console.error("NO PASSWORD");
                throw new ERRORS.BadRequest();
            };

            // TODO ADD MIDDLEWARE CHECKS
            if (req.body.password.length < 8) {
                console.error("PASSWORD < 8 CHARS");
                throw new ERRORS.BadRequest();
            };
            // =============================================================
            const PHONE_NUMBER = req.body.phone_number;
            const PASSWORD = req.body.password;
            const USER_AGENT = req.get("user-agent");

            let userId = await VERIFY_USERS(PHONE_NUMBER, PASSWORD);
            let session = await STORE_SESSION(userId, USER_AGENT);

            res.cookie("SWOB", {
                sid: session.sid,
                cookie: session.data
            }, session.data)

            return res.status(200).json({
                uid: session.uid
            });

        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(404).send(err.message);
            } // 404

            console.error(err);
            return res.status(500).send("internal server error");
        }
    });

    app.get("/users/:user_id/platforms", async (req, res, next) => {
        try {
            if (!req.params.user_id) {
                console.error("NO USERID");
                throw new ERRORS.BadRequest();
            }
            if (!req.cookies.SWOB) {
                console.error("NO COOKIE");
                throw new ERRORS.BadRequest();
            };
            const SID = req.cookies.SWOB.sid;
            const UID = req.params.user_id;
            const COOKIE = req.cookies.SWOB.cookie;
            const USER_AGENT = req.get("user-agent");

            const ID = await FIND_SESSION(SID, UID, USER_AGENT, COOKIE);
            let user = await FIND_USERS(ID);
            const usersPlatforms = await FIND_USERS_PLATFORMS(user);

            let session = await UPDATE_SESSION(SID, ID);

            res.cookie("SWOB", {
                sid: session.sid,
                cookie: session.data
            }, session.data)

            return res.status(200).json(usersPlatforms);

        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(404).send(err.message);
            } // 404

            console.error(err);
            return res.status(500).send("internal server error");
        }
    });

    app.post("/users/:user_id/platforms/:platform/protocols/:protocol", async (req, res, next) => PLATFORMS(req, res, next), async (req, res, next) => {
        // https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
        try {
            if (!req.params.user_id) {
                console.error("NO USERID");
                throw new ERRORS.BadRequest();
            }
            if (!req.cookies.SWOB) {
                console.error("NO COOKIE");
                throw new ERRORS.BadRequest();
            };
            const SID = req.cookies.SWOB.sid;
            const UID = req.params.user_id;
            const COOKIE = req.cookies.SWOB.cookie;
            const USER_AGENT = req.get("user-agent");

            const ID = await FIND_SESSION(SID, UID, USER_AGENT, COOKIE);
            const PLATFORM = req.params.platform;
            const URL = req.platformRes.url ? req.platformRes.url : "";
            const CODE = req.platformRes.code ? req.platformRes.code : "";

            let platform = await FIND_PLATFORMS(PLATFORM);

            let session = await UPDATE_SESSION(SID, ID);

            res.cookie("SWOB", {
                sid: session.sid,
                cookie: session.data
            }, session.data)

            return res.status(200).json({
                url: URL,
                code: CODE,
                platform: platform.name.toLowerCase()
            });

        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(404).send(err.message);
            } // 404

            console.error(err);
            return res.status(500).send("internal server error");
        }
    });

    app.put("/users/:user_id/platforms/:platform/protocols/:protocol/:action?", async (req, res, next) => PLATFORMS(req, res, next), async (req, res, next) => {
        try {
            if (!req.params.user_id) {
                console.error("NO USERID");
                throw new ERRORS.BadRequest();
            }
            if (!req.cookies.SWOB) {
                console.error("NO COOKIE");
                throw new ERRORS.BadRequest();
            };
            const SID = req.cookies.SWOB.sid;
            const UID = req.params.user_id;
            const COOKIE = req.cookies.SWOB.cookie;
            const USER_AGENT = req.get("user-agent");

            const ID = await FIND_SESSION(SID, UID, USER_AGENT, COOKIE);
            const PLATFORM = req.params.platform;
            const RESULT = req.platformRes.result ? req.platformRes.result : "";
            const CODE = req.platformRes.code ? req.platformRes.code : "";
            const INIT_URL = req.platformRes.initialization_url ? req.platformRes.initialization_url : "";

            let user = await FIND_USERS(ID);
            let platform = await FIND_PLATFORMS(PLATFORM);

            if (RESULT) {
                await STORE_TOKENS(user, platform, RESULT);
            }

            let session = await UPDATE_SESSION(SID, ID);

            res.cookie("SWOB", {
                sid: session.sid,
                cookie: session.data
            }, session.data)

            return res.status(200).json({
                code: CODE,
                initialization_url: INIT_URL
            });

        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(404).send(err.message);
            } // 404

            console.error(err);
            return res.status(500).send("internal server error");
        }
    });

    app.post("/users/:user_id/logout", async (req, res, next) => {
        try {
            if (!req.params.user_id) {
                console.error("NO USERID");
                throw new ERRORS.BadRequest();
            }
            if (!req.cookies.SWOB) {
                throw new ERRORS.BadRequest();
            };
            const SID = req.cookies.SWOB.sid;
            const UID = req.params.user_id;
            const COOKIE = req.cookies.SWOB.cookie;
            const USER_AGENT = req.get("user-agent");

            await FIND_SESSION(SID, UID, USER_AGENT, COOKIE);

            console.log("CLEARING BROWSER COOKIE");
            res.clearCookie("SWOB");

            return res.status(200).json();

        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(404).send(err.message);
            } // 404

            console.error(err);
            return res.status(500).send("internal server error");
        }
    });

}
// =============================================================

File: routes/user_management/v2.py
import logging
logger = logging.getLogger(__name__)

# configurations
from Configs import baseConfig
config = baseConfig()
api = config["API"]
cookie_name = api['COOKIE_NAME']
enable_otp_counter = eval(config["OTP"]["ENABLE"])

from flask import Blueprint
from flask import request
from flask import Response
from flask import jsonify

from platforms import platform_switch

from security.cookie import Cookie
from security.data import Data

import json
from datetime import timedelta

from models.grants import Grant_Model
from models.platforms import Platform_Model
from models.users import User_Model
from models.sessions import Session_Model
from models._2FA import OTP_Model

v2 = Blueprint("v2", __name__)

from werkzeug.exceptions import BadRequest
from werkzeug.exceptions import Conflict
from werkzeug.exceptions import Unauthorized
from werkzeug.exceptions import Forbidden
from werkzeug.exceptions import InternalServerError
from werkzeug.exceptions import TooManyRequests

@v2.route("/signup", methods=["POST", "PUT"])
def signup():
    """
    """
    try:
        method = request.method

        User = User_Model()
        Session = Session_Model()
        data = Data()
        cookie = Cookie()

        if method.lower() == "post":
            if not "phone_number" in request.json or not request.json["phone_number"]:
                logger.error("no phone_number")
                raise BadRequest()
            elif not "name" in request.json or not request.json["name"]:
                logger.error("no name")
                raise BadRequest()
            elif not "country_code" in request.json or not request.json["country_code"]:
                logger.error("no country_code")
                raise BadRequest()
            elif not "password" in request.json or not request.json["password"]:
                logger.error("no password")
                raise BadRequest()

            user_agent = request.headers.get("User-Agent")

            phone_number = request.json["phone_number"]
            name = request.json["name"]
            country_code = request.json["country_code"]
            password = request.json["password"]

            user_id = User.create(
                phone_number=phone_number,
                name=name,
                country_code=country_code,
                password=password
            )

            res = jsonify({
                "uid": user_id
            })

            session = Session.create(
                unique_identifier=data.hash(country_code+phone_number),
                user_agent=user_agent,
                type="signup",
            )

            cookie_data = json.dumps({
                "sid": session["sid"],
                "cookie": session["data"],
                "type":session["type"]
            })
            e_cookie = cookie.encrypt(cookie_data)

            session_data = json.loads(session["data"])

            res.set_cookie(
                cookie_name,
                e_cookie,
                max_age=timedelta(milliseconds=session_data["maxAge"]),
                secure=session_data["secure"],
                httponly=session_data["httpOnly"],
                samesite=session_data["sameSite"]
            )

        elif method.lower() == "put":
            if not request.cookies.get(cookie_name):
                logger.error("no cookie")
                raise Unauthorized()
            elif not request.headers.get("User-Agent"):
                logger.error("no user agent")
                raise BadRequest()

            e_cookie = request.cookies.get(cookie_name)
            d_cookie = cookie.decrypt(e_cookie)
            json_cookie = json.loads(d_cookie)

            sid = json_cookie["sid"]
            uid = json_cookie["uid"]
            unique_identifier = json_cookie["unique_identifier"]
            user_cookie = json_cookie["cookie"]
            type = json_cookie["type"]
            status = json_cookie["status"]
            user_agent = request.headers.get("User-Agent")

            Session.find(
                sid=sid,
                unique_identifier=unique_identifier,
                user_agent=user_agent,
                cookie=user_cookie,
                type=type,
                status=status
            )

            User.update(
                table="userinfo",
                user_id=uid,
                status="verified"
            )

            Session.update(
                sid=sid,
                unique_identifier=unique_identifier,
                status="verified",
                type=type
            )

            res = Response()

        return res, 200
                
    except BadRequest as err:
        return str(err), 400

    except Unauthorized as err:
        return str(err), 401

    except Conflict as err:
        return str(err), 409

    except InternalServerError as err:
        logger.exception(err)
        return "internal server error", 500

    except Exception as err:
        logger.exception(err)
        return "internal server error", 500

@v2.route("/users/<string:user_id>/OTP", methods=["POST"])
def OTP(user_id):
    """
    """
    try:
        if not request.cookies.get(cookie_name):
            logger.error("no cookie")
            raise Unauthorized()
        elif not request.headers.get("User-Agent"):
            logger.error("no user agent")
            raise BadRequest()
        elif not "phone_number" in request.json or not request.json["phone_number"]:
            logger.error("no phone_number")
            raise BadRequest()

        Session = Session_Model()
        data = Data()
        cookie = Cookie()

        e_cookie = request.cookies.get(cookie_name)
        d_cookie = cookie.decrypt(e_cookie)
        json_cookie = json.loads(d_cookie)

        sid = json_cookie["sid"]
        user_cookie = json_cookie["cookie"]
        type = json_cookie["type"]
        user_agent = request.headers.get("User-Agent")

        phone_number = request.json["phone_number"]

        phone_number_hash = data.hash(phone_number)
    
        Session.find(
            sid=sid,
            unique_identifier=phone_number_hash,
            user_agent=user_agent,
            cookie=user_cookie,
            type=type
        )

        otp = OTP_Model(phone_number=phone_number)

        cid = None
        expires = 0

        if enable_otp_counter:
            otp_counter = otp.check_count(
                unique_id=phone_number_hash,
                user_id=user_id
            )         

            cid = otp_counter.id

        otp_res = otp.verification()

        if otp_res.status == "pending":
            if enable_otp_counter:
                expires = otp.add_count(otp_counter)

            res = jsonify({
                "expires": expires
            })
        else:
            logger.error("OTP FAILED with status '%s'" % otp_res.status)
            raise InternalServerError(otp_res)

        session = Session.update(
            sid=sid,
            unique_identifier=phone_number_hash,
            type=type
        )

        cookie_data = json.dumps({
            "sid": session["sid"],
            "uid": user_id,
            "cookie": session["data"],
            "type": session["type"],
            "phone_number": phone_number,
            "cid": cid
        })

        e_cookie = cookie.encrypt(cookie_data)

        session_data = json.loads(session["data"])

        res.set_cookie(
            cookie_name,
            e_cookie,
            max_age=timedelta(milliseconds=session_data["maxAge"]),
            secure=session_data["secure"],
            httponly=session_data["httpOnly"],
            samesite=session_data["sameSite"]
        )

        return res, 201
                
    except BadRequest as err:
        return str(err), 400

    except Unauthorized as err:
        return str(err), 401

    except Conflict as err:
        return str(err), 409
    
    except TooManyRequests as err:
        return str(err), 429

    except InternalServerError as err:
        logger.exception(err)
        return "internal server error", 500

    except Exception as err:
        logger.exception(err)
        return "internal server error", 500

@v2.route("/OTP", methods=["PUT"])
def OTP_check():
    """
    """
    try:
        if not request.cookies.get(cookie_name):
            logger.error("no cookie")
            raise Unauthorized()
        elif not request.headers.get("User-Agent"):
            logger.error("no user agent")
            raise BadRequest()
        elif not "code" in request.json or not request.json["code"]:
            logger.error("no code")
            raise BadRequest()

        Session = Session_Model()
        data = Data()
        cookie = Cookie()

        e_cookie = request.cookies.get(cookie_name)
        d_cookie = cookie.decrypt(e_cookie)
        json_cookie = json.loads(d_cookie)

        sid = json_cookie["sid"]
        uid = json_cookie["uid"]
        user_cookie = json_cookie["cookie"]
        type = json_cookie["type"]
        phone_number = json_cookie["phone_number"]
        cid = json_cookie["cid"]
        user_agent = request.headers.get("User-Agent")

        code = request.json["code"]

        phone_number_hash = data.hash(phone_number)
    
        Session.find(
            sid=sid,
            unique_identifier=phone_number_hash,
            user_agent=user_agent,
            cookie=user_cookie,
            type=type
        )

        otp = OTP_Model(phone_number=phone_number)

        otp_res = otp.verification_check(code=code)

        if otp_res.status == "approved":
            if enable_otp_counter:
                otp.delete_count(counter_id=cid)
                
            res = Response()
        elif otp_res.status == "pending":
            logger.error("Invalid OTP code. OTP_check status = %s" % otp_res.status)
            raise Forbidden()
        else:
            logger.error("OTP_check FAILED with status '%s'" % otp_res.status)
            raise InternalServerError(otp_res)

        session = Session.update(
            sid=sid,
            unique_identifier=phone_number_hash,
            status="success",
            type=type
        )

        cookie_data = json.dumps({
            "sid": session["sid"],
            "unique_identifier": session["uid"],
            "uid": uid,
            "cookie": session["data"],
            "status": "success",
            "type": type
        })

        e_cookie = cookie.encrypt(cookie_data)

        session_data = json.loads(session["data"])

        res.set_cookie(
            cookie_name,
            e_cookie,
            max_age=timedelta(milliseconds=session_data["maxAge"]),
            secure=session_data["secure"],
            httponly=session_data["httpOnly"],
            samesite=session_data["sameSite"]
        )

        return res, 200
                
    except BadRequest as err:
        return str(err), 400

    except Unauthorized as err:
        return str(err), 401

    except Conflict as err:
        return str(err), 409

    except Forbidden as err:
        return str(err), 403

    except InternalServerError as err:
        logger.exception(err)
        return "internal server error", 500

    except Exception as err:
        logger.exception(err)
        return "internal server error", 500

@v2.route("users/<string:user_id>/platforms/<string:platform>/protocols/<string:protocol>/", defaults={"action": None}, methods=["POST", "PUT", "DELETE"])
@v2.route("users/<string:user_id>/platforms/<string:platform>/protocols/<string:protocol>/<string:action>", methods=["PUT"])
async def manage_grant(user_id, platform, protocol, action) -> dict:
    """
    """
    try:
        originalUrl = request.host_url
        method = request.method

        if not "code" in request.json or not request.json["code"]:
            code = None
        else:
            code = request.json["code"]
        
        if not "phone_number" in request.json or not request.json["phone_number"]:
            phone_number = None
        else:
            phone_number = request.json["phone_number"]

        if not "code_verifier" in request.json or not request.json["code_verifier"]:
            code_verifier = None
        else:
            code_verifier = request.json["code_verifier"]

        if not "first_name" in request.json or not request.json["first_name"]:
            first_name = None
        else:
            first_name = request.json["first_name"]
        
        if not "last_name" in request.json or not request.json["last_name"]:
            last_name = None
        else:
            last_name = request.json["last_name"]

        Platform = Platform_Model()
        Grant = Grant_Model()

        if method.lower() == "post":
            result = await platform_switch(
                originalUrl=originalUrl,
                platform_name=platform,
                protocol=protocol,
                method=method,
                phoneNumber=phone_number
            )

            if not "url" in result or not result["url"]:
                url = ""
            else:
                url = result["url"]

            if not "code_verifier" in result or not result["code_verifier"]:
                code_verifier = ""
            else:
                code_verifier = result["code_verifier"]

            if not "body" in result or not result["body"]:
                body = ""
            else:
                body = result["body"]

            res = jsonify({
                "url": url,
                "code_verifier": code_verifier,
                "body": body,
                "platform": platform.lower()
            })

        elif method.lower() == "put":                     
            result = await platform_switch(
                originalUrl=originalUrl,
                platform_name=platform,
                protocol=protocol,
                method=method,
                code=code,
                code_verifier=code_verifier,
                phoneNumber=phone_number,
                action=action,
                first_name=first_name,
                last_name=last_name
            )

            if not "body" in result or not result["body"]:
                body = ""
            else:
                body = result["body"]

            if not "initialization_url" in result or not result["initialization_url"]:
                initialization_url = ""
            else:
                initialization_url = result["initialization_url"]

            if not "grant" in result or not result["grant"]:
                grant = None
            else:
                grant = result["grant"]

            platform_result = Platform.find(platform_name=platform)

            if grant:
                Grant.store(
                    user_id=user_id,
                    platform_id=platform_result.id,
                    platform_name=platform_result.name,
                    grant=grant
                )
          
            res = jsonify({
                "body": body,
                "initialization_url": initialization_url
            })

        elif method.lower() == "delete":
            platform_result = Platform.find(platform_name=platform)
            grant = Grant.find(user_id=user_id, platform_id=platform_result.id)

            result = await platform_switch(
                originalUrl=originalUrl,
                platform_name=platform,
                protocol=protocol,
                method=method,
                phoneNumber=grant.token,
                token=grant.token
            )

            Grant.delete(grant=grant)

            res = Response()
        
        return res, 200

    except BadRequest as error:
        return str(error), 400

    except Conflict as error:
        return str(error), 409

    except InternalServerError as error:
        logger.exception(error)
        return "internal server error", 500

    except Exception as error:
        logger.exception(error)
        return "internal server error", 500